netbox/dcim/filtersets.py

```diff
@@ -981,7 +981,9 @@ def search(self, queryset, name, value):
             Q(serial__icontains=value.strip()) |
             Q(inventoryitems__serial__icontains=value.strip()) |
             Q(asset_tag__icontains=value.strip()) |
-            Q(comments__icontains=value)
+            Q(comments__icontains=value) |
+            Q(primary_ip4__address__startswith=value) |
+            Q(primary_ip6__address__startswith=value)
```

Comment:
I don't have any strong feelings about this. There's a performance penalty (on my test system, ~10k devices and ~14k interfaces the query time goes from ~45 ms to ~70 ms. However I can see how it could be very useful and since the q search is no longer part of the global search, the performance penalty might be worth it?

Reply:
It is indeed quite a hit.. Did not test that... but still acceptable for the extra searchability of a device.
I often get an IP where I have to find the device for, this will make that search much more direct

Reply:
> but still acceptable for the extra searchability of a device.

We don't really have any way of knowing this. A user with 100K IPs defined is going to experience a _far_ more severe performance hit than someone with a few hundred (hence my comment [here](https://github.com/netbox-community/netbox/issues/10058#issuecomment-1222922659)).

IMO we should evaluate potential strategies for caching the IP addresses directly on the device model rather than imposing a potentially substantial burden on the user.

Reply:
Another consideration: It might be more efficient to perform a separate query against IPAddress (for relevant search queries), and then match on the ID for any device that points to matching IPAddresses. This could be a union with devices matching on other fields.

Additionally, `startswith` probably isn't a sufficient lookup for IP addresses as it's constrained to matching on the octet/nibble boundary,

Reply:
The search is on the primary ip assigned to a device, not to the list of IPs.
And this is for the quick search/device search, not for the new cached search.

Reply:
> The search is on the primary ip assigned to a device, not to the list of IPs.

As it's currently constructed, the query is triggering a SQL `JOIN` with the IP addresses table (two actually), which is very expensive. I'm saying we could instead query the IP addresses table directly for the value, and then work backward to device the IDs of any relevant devices:

1. Does the search term look like an IP address? (If no, skip these steps.)
2. Find all matching IP addresses.
3. For all matching IPs, find all devices pointing to them as primary IPs.
4. Add the IDs of these devices to the base query.

Reply:
I built a test bed for this with 10 thousand devices, each with an interface and both IPv4 and IPv6 primary IPs. The resulting SQL query when searching for e.g. `?q=10.1.2.3` takes only 40-50ms on average, which seems very reasonable and probably wouldn't benefit substantially from further optimization. I'm open to potential future work in the event of complaints, but this seems acceptable as-is. Thanks for your work on this @PieterL75!